sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java

```diff
@@ -149,9 +152,54 @@ public ReadableByteChannel createDecompressingChannel(ReadableByteChannel channe
         return Channels.newChannel(
             new BZip2CompressorInputStream(Channels.newInputStream(channel)));
       }
+    },
+
+    /**
+     * Reads a byte channel assuming it is compressed with zip.
+     * If the zip file contains multiple entries, files in the zip are concatenated all together.
+     */
+    ZIP {
+      @Override
+      public boolean matches(String fileName) {
+        return fileName.toLowerCase().endsWith(".zip");
+      }
+
+      public ReadableByteChannel createDecompressingChannel(ReadableByteChannel channel)
+        throws IOException {
+        FullZipInputStream zip = new FullZipInputStream(Channels.newInputStream(channel));
+        return Channels.newChannel(zip);
+      }
     };
 
     /**
+     * Extend of {@link ZipInputStream} to automatically read all entries in the zip.
+     */
+    public class FullZipInputStream extends ZipInputStream {
+
+      private ZipEntry currentEntry;
+
+      public FullZipInputStream(InputStream is) throws IOException {
+        super(is);
+        currentEntry = getNextEntry();
+      }
+
+      public int read(byte[] b, int off, int len) throws IOException {
+        int result = super.read(b, off, len);
+        if (result == -1) {
+          currentEntry = getNextEntry();
+          if (currentEntry == null) {
+            return -1;
+          }
+          result = super.read(b, off, len);
+          return result;
```

Comment:
I admit two empty files is relatively unlikely, though totally possible. But this is not just some weird corner case--there are other execution paths in which it is possible for the stream to read exactly to the end of a file but not hit EOF (so `-1` is not returned yet), and then it will act like an empty file for the purposes of the next call into the function.